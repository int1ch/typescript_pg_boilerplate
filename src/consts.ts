export const SMS_NOTIFICATION = "sms_notification";
export const EMAIL_NOTIFICATION = "email_notification";

export const UserConsentTypeVariants = [
  SMS_NOTIFICATION,
  EMAIL_NOTIFICATION,
] as const;
export type UserConsentType = typeof UserConsentTypeVariants[number];
