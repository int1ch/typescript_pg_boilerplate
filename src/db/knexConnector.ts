import { knex } from "knex";
import { snakecase, camelcase } from "stringcase";
import * as config from "../config";
import logger from "../logger";

export async function testConnection() {
  await connection.raw("SELECT now()");
  return connection;
}

const DUMP = false;
const DEBUG = false;

export function openConnection() {
  //ups, dev may run in testing / development
  //other envs expected to use only PG_PASSWORD,
  const password = config.IS_TESTING
    ? config.PG_TESTING_PASSWORD
    : config.PG_PASSWORD;

  const connection = knex({
    client: "pg",
    connection: {
      host: config.PG_HOST,
      port: config.PG_PORT,
      user: config.PG_USERNAME,
      password: password,
      database: config.PG_DATABASE,
    },
    acquireConnectionTimeout: 3000,
    pool: {
      min: 2,
      max: 2, //FIXME
    },
    wrapIdentifier: snakeWrapper,
    postProcessResponse: postConverter,
    debug: DEBUG,
  });
  return connection;
}

function snakeWrapper(
  value: string,
  origImpl: (value: string) => string,
  queryContext: any
): string {
  if (DUMP) logger.log("snake wrapper:", value, origImpl, queryContext);

  return origImpl(snakecase(value));
}
function postConverter(value: any, queryContext: any, path?: string): any {
  if (isRaw(value, queryContext, path)) return value;
  if (DUMP) logger.log("Query POST Converter:", value);
  if (Array.isArray(value)) {
    return value.map((item) => postConverter(item, queryContext));
  }
  if (value._parsers && value._types) {
    const output: any = {};
    Object.assign(output, value);
    output.rows = postConverter(value.rows, queryContext);
    return output;
  }

  const output: any = {};
  path ||= "";
  for (const key of Object.keys(value)) {
    const newKey = camelcase(key);
    output[newKey] = value[key];
    /*postConverter(
      value[key],
      queryContext,
      (path += "." + key)
    );*/
  }

  return output;
}

function isRaw(value: any, queryContext: any, path?: string) {
  if (typeof value !== "object" || value === null) return true;
  if (value instanceof Date) return true;
  return false;

  //if (path === "root") return false;
  //if (typeof recursive !== "function") return true;
  //return !recursive(value, path, queryContext);
}

export async function closeConnection() {
  connection.destroy();
  //logger.log("unsupported?");
  //connection.close
}
const connection = openConnection();
export default connection;
