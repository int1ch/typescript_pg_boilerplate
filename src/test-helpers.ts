import { userInfo } from "os";
import dbUser, { DbUser } from "./db/User";

export function fixUuidIn(body: any, field: string) {
  try {
    body[field] = "uuid0000-fixe-df04-test-reason6922cd";
  } catch (e) {}
}

export function fixEmailIn(body: any, field: string) {
  try {
    body[field] = "fixed@email.ru";
  } catch (e) {}
}

function randInt(max = 100000) {
  Math.ceil(Math.random() * 10000);
}

export class UserCreator {
  protected dbUser: DbUser;
  protected createdIds: string[];
  protected n = 0;
  constructor() {
    this.dbUser = dbUser;
    this.createdIds = [];
  }
  async create(options?: { email?: string }) {
    const rand = randInt();
    const email = options?.email || `cr-${this.n++}-${rand}@test-cr.com`;

    const userId = await this.dbUser.create({ email: email });
    this.createdIds.push(userId);
    return userId;
  }
  async cleanUp() {
    for (const userId of this.createdIds) {
      await this.dbUser.delete(userId);
    }
  }
}

export const userCreatorHelper = new UserCreator();
