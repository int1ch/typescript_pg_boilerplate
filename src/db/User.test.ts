import { EMAIL_NOTIFICATION, SMS_NOTIFICATION } from "../consts";
import { fixUuidIn } from "../test-helpers";
import { closeConnection } from "./knexConnector";
import dbUser, { UserConsent } from "./User";

//@ts-ignore
const knex = dbUser.knex;

const email1 = "one@email.com";
const email2 = "two@mail.ru";
const email3 = "three@gmail.ru";
const email_for_chaenge = "changed@hotmail.com";
const emails = [email1, email2, email3];

afterAll(async () => {
  await closeConnection();
});
beforeAll(async () => {
  await knex("users")
    .whereIn("email", [...emails, email_for_chaenge])
    .delete();
});
const uuids: string[] = [];
describe("users", () => {
  it("create", async () => {
    for (const email of emails) {
      const uuid = await dbUser.create({ email: email });
      expect(uuid).toHaveLength(36);
      uuids.push(uuid);
    }
  });

  it("double create protection", async () => {
    await expect(async () => {
      await dbUser.create({ email: email1 });
    }).rejects.toThrowError("Email one@email.com already used by some one");
  });
  it("update email protection", async () => {
    await expect(async () => {
      await dbUser.update({ userId: uuids[1], email: email1 });
    }).rejects.toThrowError("Email one@email.com already used by some one");
  });

  it("get normally", async () => {
    const user = await dbUser.get(uuids[0]);
    //expect(user);
    expect(user!.userId).toEqual(uuids[0]);
    expect(user!.email).toEqual(emails[0]);
  });

  it("update email ok", async () => {
    dbUser.update({ userId: uuids[1], email: email_for_chaenge });
    const user = await dbUser.get(uuids[1]);
    //expect(user);
    expect(user!.email).toEqual(email_for_chaenge);
  });

  describe("users consent", () => {
    it("set consent", async () => {
      await dbUser.setUserConsent({
        userId: uuids[1],
        type: EMAIL_NOTIFICATION,
        enabled: true,
      });

      const consents = await dbUser.getUserConsents(uuids[1]);
      expect(consents).toHaveLength(1);
      fixUuidInConsents(consents);
      expect(consents).toMatchSnapshot();

      const consentsHistory = await dbUser.getUserConsentsHistory(uuids[1]);
      expect(consentsHistory).toHaveLength(1);
    });
    it("set second consent", async () => {
      await dbUser.setUserConsent({
        userId: uuids[1],
        type: SMS_NOTIFICATION,
        enabled: true,
      });

      const consents = await dbUser.getUserConsents(uuids[1]);
      expect(consents).toHaveLength(2);
      fixUuidInConsents(consents);
      expect(consents).toMatchSnapshot();

      const consentsHistory = await dbUser.getUserConsentsHistory(uuids[1]);
      expect(consentsHistory).toHaveLength(2);
    });
    it("rewrite consent", async () => {
      await dbUser.setUserConsent({
        userId: uuids[1],
        type: SMS_NOTIFICATION,
        enabled: false,
      });

      const consents = await dbUser.getUserConsents(uuids[1]);
      expect(consents).toHaveLength(2);
      fixUuidInConsents(consents);
      expect(consents).toMatchSnapshot();

      const consentsHistory = await dbUser.getUserConsentsHistory(uuids[1]);
      expect(consentsHistory).toHaveLength(3);
      const lastEvent = consentsHistory[0];
      expect(lastEvent.type).toBe(SMS_NOTIFICATION);
      expect(lastEvent.enabled).toBe(false);
    });
  });
});

function fixUuidInConsents(consents: UserConsent[]) {
  for (const consent of consents) {
    fixUuidIn(consent, "userId");
  }
  return consents;
}
