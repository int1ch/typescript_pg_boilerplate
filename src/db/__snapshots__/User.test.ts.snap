// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users users consent rewrite consent 1`] = `
Array [
  Object {
    "enabled": true,
    "type": "email_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
  Object {
    "enabled": false,
    "type": "sms_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
]
`;

exports[`users users consent set consent 1`] = `
Array [
  Object {
    "enabled": true,
    "type": "email_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
]
`;

exports[`users users consent set second consent 1`] = `
Array [
  Object {
    "enabled": true,
    "type": "email_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
  Object {
    "enabled": true,
    "type": "sms_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
]
`;
